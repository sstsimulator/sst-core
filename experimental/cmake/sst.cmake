# Set variables for sst_config.h attempt to just go in order
include(CheckSymbolExists)
include(CheckIncludeFile)
include(CheckLibraryExists)

check_symbol_exists(argz_add "argz.h" HAVE_ARGZ_ADD)
check_symbol_exists(argz_append "argz.h" HAVE_ARGZ_APPEND)
check_symbol_exists(argz_count "argz.h" HAVE_ARGZ_COUNT)
check_symbol_exists(argz_create_sep "argz.h" HAVE_ARGZ_CREATE_SEP)
check_include_file(argz.h HAVE_ARGZ_H)
check_symbol_exists(argz_insert "argz.h" HAVE_ARGZ_INSERT)
check_symbol_exists(argz_next "argz.h" HAVE_ARGZ_NEXT)
check_symbol_exists(argz_stringify "argz.h" HAVE_ARGZ_STRINGIFY)
check_symbol_exists(closedir "sys/types.h" HAVE_CLOSEDIR)
check_include_file(c_asm.h HAVE_C_ASM_H)
check_symbol_exists(cygwin_conv_path "sys/cygwin.h" HAVE_DECL_CYGWIN_CONV_PATH)
check_include_file(dirent.h HAVE_DIRENT_H)
check_library_exists(dld dld_init "" HAVE_DLD)
check_include_file(dld.h HAVE_DLD_H)
check_symbol_exists(dlerror "dlfcn.h" HAVE_DLERROR)
check_include_file(dlfcn.h HAVE_DLFCN_H)
check_include_file(dl.h HAVE_DL_H)
check_symbol_exists(_dyld_func_lookup "dyld.h" HAVE_DYLD)
check_symbol_exists(error_t "errno.h" HAVE_ERROR_T)

if(HDF5_FOUND)
  set(HAVE_HDF5 ON)
endif(HDF5_FOUND)

check_include_file(intrinsics.h HAVE_INTRINSICS_H)
check_include_file(inttypes.h HAVE_INTTYPES_H)
if(HAVE_INTTYPES_H)
  # TODO do we need to check that symbols like PRIu64 exist first?
  set(__STDC_FORMAT_MACROS ON)
endif(HAVE_INTTYPES_H)

if(CMAKE_DL_LIBS)
  set(HAVE_LIBDL ON)
endif()

if(ZLIB_FOUND)
  set(HAVE_LIBZ ON)
endif(ZLIB_FOUND)

if(CURSES_FOUND)
  set(HAVE_CURSES ON)
endif(CURSES_FOUND)

check_include_file(execinfo.h HAVE_EXECINFO_H)
check_symbol_exists(backtrace "execinfo.h" HAVE_BACKTRACE)
check_include_file(mach/mach_time.h HAVE_MACH_MACH_TIME_H)
check_include_file(mach-o/dyld.h HAVE_MACH_O_DYLD_H)
check_symbol_exists(opendir "dirent.h" HAVE_OPENDIR)
check_symbol_exists(readdir "dirent.h" HAVE_READDIR)

# TODO we really should fix how we do c++ std flags
if(CMAKE_CXX_STANDARD GREATER_EQUAL 14)
  set(HAVE_STDCXX_1Y ON)
endif()

check_include_file(stdint.h HAVE_STDINT_H)
check_include_file(stdio.h HAVE_STDIO_H)
check_include_file(stdlib.h HAVE_STDLIB_H)
check_include_file(strings.h HAVE_STRINGS_H)
check_include_file(string.h HAVE_STRING_H)
check_symbol_exists(strlcat "string.h" HAVE_STRLCAT)
check_symbol_exists(strlcpy "string.h" HAVE_STRLCPY)
check_include_file(sys/dl.h HAVE_SYS_DL_H)
check_include_file(sys/stat.h HAVE_SYS_STAT_H)
check_include_file(sys/stat.h HAVE_SYS_STAT_H)
check_include_file(sys/time.h HAVE_SYS_TIME_H)
check_include_file(sys/types.h HAVE_SYS_TYPES_H)
check_include_file(unistd.h HAVE_UNISTD_H)

if(HAVE_ARGZ_H)
  set(HAVE_WORKING_ARGZ ON)
endif()

set(PACKAGE_BUGREPORT "https://github.com/sstsimulator/sst-core/issues")
set(PACKAGE_NAME "SSTCore")
set(PACKAGE_NAME "https://github.com/sstsimulator/sst-core")

set(SST_CC ${CMAKE_C_COMPILER})
set(SST_CFLAGS ${CMAKE_C_FLAGS})

if(APPLE)
  set(SST_COMPILE_MACOSX 1)
endif()

if(MPI_FOUND)
  set(SST_CONFIG_HAVE_MPI ON)
  set(SST_MPICC ${MPI_C_COMPILER})
  set(SST_MPICXX ${MPI_CXX_COMPILER})
endif()

if(Python_FOUND)
  set(HAVE_PYTHON_H 1)
  set(SST_CONFIG_HAVE_PYTHON ON)
  # CMAKE build can just require PYTHON3
  set(SST_CONFIG_HAVE_PYTHON3 ON)
  set(SST_PTYHON_LDFLAGS "${Python_LIBRARIES}")
  set(SST_PYTHON_CPPFLAGS "-I${Python_INCLUDE_DIRS}")

  # NOTE I think we should try to avoid using python-config if possible
  # find_program(PYCONFIG NAMES python-config "${Python_EXECUTABLE}-config")
  # if(PYCONFIG) message(STATUS "Found PYCONFIG ${PYCONFIG}.")
  #
  # #execute_process( #  COMMAND ${PYCONFIG} --ldflags #  RESULT_VARIABLE
  # PYCONFIG_RESULT #  OUTPUT_VARIABLE SST_PYTHON_LDFLAGS #
  # OUTPUT_STRIP_TRAILING_WHITESPACE)
  #
  # #if(PYCONFIG_RESULT AND NOT PYCONFIG_RESULT EQUAL 0) #  message( #
  # FATAL_ERROR #      "python-config (${PYCONFIG}) --ldflags failed with
  # ${PYCONFIG_RESULT}") #endif(PYCONFIG_RESULT AND NOT PYCONFIG_RESULT EQUAL 0)
  # set(SST_PTYHON_LDFLAGS ${Python_LIBRARIES})
  #
  # execute_process( COMMAND ${PYCONFIG} --cppflags RESULT_VARIABLE
  # PYCONFIG_RESULT OUTPUT_VARIABLE SST_PYTHON_CPPFLAGS
  # OUTPUT_STRIP_TRAILING_WHITESPACE) else(PYCONFIG) message(FATAL_ERROR "Failed
  # to find python-config") endif(PYCONFIG)
endif()

set(SST_CPP ${CMAKE_CXX_COMPILER})
set(SST_CPPFLAGS ${CMAKE_CXX_FLAGS})
set(SST_CXXCPP ${CMAKE_CXX_COMPILER})
set(SST_CXX ${CMAKE_CXX_COMPILER})
set(SST_CXXFLAGS ${CMAKE_CXX_FLAGS})
set(SST_INSTALL_PREFIX ${CMAKE_INSTALL_PREFIX})
set(SST_LD ${CMAKE_LINKER})
set(SST_LDFLAGS "${CMAKE_CXX_LINK_FLAGS} ${LINK_FLAGS}")

# This is a workaround for sstsimulator.conf LD, the autotools expects LIBS to
# be -l lib but if CMAKE_DL_LIBS is empty you just get a -l whic breaks linking
if(CMAKE_DL_LIBS)
  set(SST_DL_LIBS "-l${CMAKE_DL_LIBS}")
endif()

if(NOT SST_DISABLE_MEM_POOLS)
  set(USE_MEMPOOL ON)
endif()

set(SST_BUILD_WITH_CMAKE ON)
set(PACKAGE_VERSION ${CMAKE_PROJECT_VERSION})

if(Python_VERSION AND Python_VERSION_MAJOR GREATER_EQUAL 3)
  set(SST_CONFIG_HAVE_PYTHON3 ON)
  set(HAVE_PYTHON_H ON)
endif()

find_package(Git REQUIRED)
execute_process(
  COMMAND ${GIT_EXECUTABLE} --git-dir=${SST_TOP_SRC_DIR}/.git rev-parse HEAD
  RESULT_VARIABLE HASH_RESULT
  OUTPUT_VARIABLE SSTCORE_GIT_HEADSHA
  OUTPUT_STRIP_TRAILING_WHITESPACE)

execute_process(
  COMMAND ${GIT_EXECUTABLE} --git-dir=${SST_TOP_SRC_DIR}/.git rev-parse
          --abbrev-ref HEAD
  RESULT_VARIABLE BRANCH_RESULT
  OUTPUT_VARIABLE SSTCORE_GIT_BRANCH
  OUTPUT_STRIP_TRAILING_WHITESPACE)

check_include_file(sys/stat.h HAVE_SYS_STAT_H)
check_include_file(dlfcn.h HAVE_DLFCN_H)
check_include_file(sys/types HAVE_SYS_TYPES_H)
check_include_file(unistd.h HAVE_UNISTD_H)

check_library_exists(m sin "" HAVE_LIBM)
if(NOT HAVE_LIBM)
  message(FATAL_ERROR "Failed to detect libm")
endif()

check_library_exists(rt aio_cancel "" HAVE_LIBRT)
