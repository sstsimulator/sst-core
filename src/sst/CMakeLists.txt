# ~~~
# SST-CORE src/sst/ CMake
#
# Copyright 2009-2023 NTESS. Under the terms
# of Contract DE-NA0003525 with NTESS, the U.S.
# Government retains certain rights in this software.
#
# Copyright (c) 2009-2023, NTESS
# All rights reserved.
#
# This file is part of the SST software package. For license
# information, see the LICENSE file in the top level directory of the
# distribution.
# ~~~
#

add_subdirectory(core)
