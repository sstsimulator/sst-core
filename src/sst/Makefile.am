# Copyright 2009-2024 NTESS. Under the terms
# of Contract DE-NA0003525 with NTESS, the U.S.
# Government retains certain rights in this software.
#
# Copyright (c) 2009-2024, NTESS
# All rights reserved.
#
# This file is part of the SST software package. For license
# information, see the LICENSE file in the top level directory of the
# distribution.

SUBDIRS = \
	core

pkgconfigdir = $(libdir)/pkgconfig
pkgconfig_DATA = SST-${PACKAGE_VERSION}.pc
