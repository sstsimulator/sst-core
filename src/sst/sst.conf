# ===================================================================
# SST Configuration Database
#
# This file is automatically generated during installation. Any
# changes made manually may be overridden. Please see the SST
# documentation for details on how the configurations should be
# edited to prevent errors during execution
#
# ===================================================================
[SSTCore]
MPICXX=@MPICXX@
MPICC=@MPICC@
MPI_CPPFLAGS=@MPI_CPPFLAGS@
CXX=@CXX@
CXXFLAGS=@SST_EXPORT_CXXFLAGS@
CC=@CC@
CFLAGS=@SST_EXPORT_CFLAGS@
LD=@LD@
LDFLAGS=@LDFLAGS@
LIBS=@LIBS@
CPP=@CPP@
CXXCPP=@CXXCPP@
AR=@AR@
NM=@NM@
PYTHON=@PYTHON_EXE@
PYTHON_CONFIG=@PYTHON_CONFIG_EXE@
PYTHON_CPPFLAGS=@PYTHON_CPPFLAGS@
PYTHON_LDFLAGS=@PYTHON_LDFLAGS@
PYTHON_VERSION3=@PYTHON_VERSION3@
ZOLTAN_CPPFLAGS=@ZOLTAN_CPPFLAGS@
ZOLTAN_LDFLAGS=@ZOLTAN_LDFLAGS@
LIBZ_CPPFLAGS=@LIBZ_CPPFLAGS@
LIBZ_LDFLAGS=@LIBZ_LDFLAGS@
LIBZ_LIBS=@LIBZ_LIBS@
ELEMENT_CXXFLAGS=@SST_EXPORT_CXXFLAGS@ @SST_ELEMENT_FPIC_FLAGS@ -DHAVE_CONFIG_H -I@prefix@/include
ELEMENT_LDFLAGS=-shared -fno-common -Wl,-undefined -Wl,dynamic_lookup
pkgconfig=@prefix@/lib/pkgconfig/SST-@PACKAGE_VERSION@.pc
sourcedir=@abs_srcdir@/core
testsdir=@abs_srcdir@/../../tests
prefix=@prefix@
exec_prefix=@prefix@
bindir=@prefix@/bin
sbindir=@prefix@/sbin
libexecdir=@prefix@/libexec
includedir=@prefix@/include
libdir=@prefix@/lib
CORE_TEST_ELEMENT_LIBDIR=@prefix@/lib/sstcore
PREVIEW_BUILD=@SST_ENABLE_PREVIEW_BUILD_STR@
