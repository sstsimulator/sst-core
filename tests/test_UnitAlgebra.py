from sst import UnitAlgebra


ua1 = UnitAlgebra("15ns")
ua2 = UnitAlgebra("10ns")

# Test all the math functions
print("Simple math operations:")

#addition
ua3 = ua1 + ua2
print("%s + %s = %s"%(ua1, ua2, ua3))

#subtraction
ua3 = ua1 - ua2
print("%s - %s = %s"%(ua1, ua2, ua3))

#multiplication
ua3 = ua1 * ua2
print("%s * %s = %s"%(ua1, ua2, ua3))

#division
ua3 = ua1 / ua2
print("%s / %s = %s"%(ua1, ua2, ua3))


# in-place functions
print("")
print("In-place math operations:")
print("executing the following for each operator:")
print("ua1 = UnitAlgebra(\"15ns\")")
print("ua2 = UnitAlgebra(\"10ns\")")
print("ua3 = ua1")
print("ua4 = ua3")
print("ua3 op= ua2")
print("print(ua3)  -- should have the result of operation")
print("print(ua4)  -- should be unchanged")
print("")

# During this, test that assignment and in-place operators work
# correctly as far as return new objects

# a = b should make a and b the same object.

# funny enough, a += b will return a new object for a (this is how
# it's supposed to work

print("Addition:")

ua3 = ua1
if id(ua3) != id(ua1):
    print("ERROR: assignement operator did not result in both objects being the same")
ua4 = ua3
ua3_id_before = id(ua3)
ua3 += ua2
if id(ua3) == ua3_id_before:
    print("ERROR: += operator returned the same object")
print(ua3)
print(ua4)

print("Subtraction:")

ua3 = ua1
ua4 = ua3
ua3_id_before = id(ua3)
ua3 -= ua2
if id(ua3) == ua3_id_before:
    print("ERROR: -= operator returned the same object")
print(ua3)
print(ua4)

print("Multiplication:")

ua3 = ua1
ua4 = ua3
ua3_id_before = id(ua3)
ua3 *= ua2
if id(ua3) == ua3_id_before:
    print("ERROR: *= operator returned the same object")
print(ua3)
print(ua4)

print("Division:")

ua3 = ua1
ua4 = ua3
ua3_id_before = id(ua3)
ua3 /= ua2
if id(ua3) == ua3_id_before:
    print("ERROR: \= operator returned the same object")
print(ua3)
print(ua4)

# Check to see if exceptions are triggered as they should be if the
# operands aren't correct

# Only need to try one math operation since they all call into the
# same underlying function where the type checks happen
correct_throw = False
try:
    ua1 = UnitAlgebra("10ns")
    ua2 = ua1 + 5
except TypeError:
    print("Exception correctly thrown for wrong operand type passed to math operation (+,-,*,/) as second argument")
    correct_throw = True

if not correct_throw:
    print("ERROR: Exception NOT correctly thrown for wrong operand type passed to math operation (+,-,*,/) as second argument")


correct_throw = False
try:
    ua1 = UnitAlgebra("10ns")
    ua2 = 5 + ua1
except TypeError:
    print("Exception correctly thrown for wrong operand type passed to math operation (+,-,*,/) as first argument")
    correct_throw = True

if not correct_throw:
    print("ERROR: Exception NOT correctly thrown for wrong operand type passed to math operation (+,-,*,/) as first argument")



print("")

print("Comparison functions:")
ua3 = UnitAlgebra("15ns")

print("%s > %s = %r"%(ua1,ua2,ua1 > ua2))
print("%s >= %s = %r"%(ua1,ua2,ua1 >= ua2))
print("%s < %s = %r"%(ua1,ua2,ua1 < ua2))
print("%s <= %s = %r"%(ua1,ua2,ua1 <= ua2))
print("%s == %s = %r"%(ua1,ua2,ua1 == ua2))
print("%s != %s = %r"%(ua1,ua2,ua1 != ua2))
print("")
print("%s > %s = %r"%(ua1,ua3,ua1 > ua3))
print("%s >= %s = %r"%(ua1,ua3,ua1 >= ua3))
print("%s < %s = %r"%(ua1,ua3,ua1 < ua3))
print("%s <= %s = %r"%(ua1,ua3,ua1 <= ua3))
print("%s == %s = %r"%(ua1,ua3,ua1 == ua3))
print("%s != %s = %r"%(ua1,ua3,ua1 != ua3))


# Check to make sure we get exceptions for invalid arguments.  Since
# it is the same function, really only need to check one operator.
# Need to check using the UnitAlgebra as both the first and second
# argument.  Also need to check equals and not equals separately,
# because they can fall back on just checking to see if they are the
# exact same object, in which case, then won't generate an error.
print("")
correct_throw = False
try:
    ua1 > 5
except TypeError:
    correct_throw = True
    print("Correctly got exception when comparing against invalid operand type as second argument")

if not correct_throw:
    print("ERROR: did not get exception when comparing against invalid argument type as second argument")

correct_throw = False
try:
    5 > ua1
except TypeError:
    correct_throw = True
    print("Correctly got exception when comparing against invalid operand type as first argument")

if not correct_throw:
    print("ERROR: did not get exception when comparing against invalid argument type as first argument")


correct_throw = False
try:
    ua1 == 5
except TypeError:
    correct_throw = True
    print("Correctly got exception when checking equals against invalid operand type as second argument")

if not correct_throw:
    print("ERROR: did not get exception when comparing against invalid operand type as second argument")


correct_throw = False
try:
    5 == ua1
except TypeError:
    correct_throw = True
    print("Correctly got exception when checking equals against invalid operand type as first argument")

if not correct_throw:
    print("ERROR: did not get exception when checking equals against invalid operand type as first argument")


correct_throw = False
try:
    ua1 != 5
except TypeError:
    correct_throw = True
    print("Correctly got exception when checking not equals against invalid operand type as second argument")

if not correct_throw:
    print("ERROR: did not get exception when checking not equals against invalid operand type as second argument")

correct_throw = False
try:
    5 != ua1
except TypeError:
    correct_throw = True
    print("Correctly got exception when checking not equals against invalid operand type as first argument")

if not correct_throw:
    print("ERROR: did not get exception when checking not equals against invalid operand type as first argument")


# Once expections are enabled in UnitAlgebra (instead of the fatal()
# that is currently called), we need to test doing math and comparison
# operations on this with mismatch units to make sure that they throw
# exceptions in the proper instances (i.e. add, sub and compares other
# than == and != will throw an exception if the units don't match.



#Check conversions to int (in which case it rounds to nearest integer)
#and floats.  There are two ways to do it:
# 1 - use the built in int() and float() operations
# 2 - use the getRoundedValue() and getFloatValue() functions
#     defined on UnitAlgebra
print("")
print("Conversion to int:")
ua3 = UnitAlgebra("1GHz")
print("'%s' to long = %d"%(ua3,int(ua3)))
print("'%s'.getRoundedValue() = %d"%(ua3, ua3.getRoundedValue()))

print("")
print("Conversion to float:")
ua3 = UnitAlgebra("1.77s")
print("'%s' to float = %f"%(ua3,float(ua3)))
print("'%s'.getFloatValue() = %f"%(ua3, ua3.getFloatValue()))

print("")
print("Conversion to float:")
ua3 = UnitAlgebra("1.77s")
print("'%s' to float = %f"%(ua3,float(ua3)))
print("'%s'.getFloatValue() = %f"%(ua3, ua3.getFloatValue()))

print("")
print("Conversion to bool:")
ua3 = UnitAlgebra("0ns")
print("bool(%s) = %r"%(ua3,bool(ua3)))
ua3 = UnitAlgebra("5ns")
print("bool(%s) = %r"%(ua3,bool(ua3)))

# Test the remaining functions that haven't been tested elsewhere
print("")
print("isValueZero():")
ua3 = UnitAlgebra("0ns")
print("'%s'.isValueZero() = %r"%(ua3,ua3.isValueZero()))
ua3 = UnitAlgebra("5ns")
print("'%s'.isValueZero() = %r"%(ua3,ua3.isValueZero()))

print("")
print("Negate:")
print("-%s = %s"%(ua1,-ua1))

print("")
print("Invert:")
print("'%s'.invert() = %s"%(ua1,ua1.invert()))

print("")
print("hasUnits():")
print("'%s'.hasUnits(\"ns\") = %r"%(ua1,ua1.hasUnits("ns")))
print("'%s'.hasUnits(\"s\") = %r"%(ua1,ua1.hasUnits("s")))
print("'%s'.hasUnits(\"Hz\") = %r"%(ua1,ua1.hasUnits("Hz")))

