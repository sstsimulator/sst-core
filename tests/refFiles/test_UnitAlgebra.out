Simple math operations:
15 ns + 10 ns = 25 ns
15 ns - 10 ns = 5 ns
15 ns * 10 ns = 150 as-s
15 ns / 10 ns = 1.5 

In-place math operations:
executing the following for each operator:
ua1 = UnitAlgebra("15ns")
ua2 = UnitAlgebra("10ns")
ua3 = ua1
ua4 = ua3
ua3 op= ua2
print(ua3)  -- should have the result of operation
print(ua4)  -- should be unchanged

Addition:
25 ns
15 ns
Subtraction:
5 ns
15 ns
Multiplication:
150 as-s
15 ns
Division:
1.5 
15 ns
Exception correctly thrown for wrong operand type passed to math operation (+,-,*,/) as second argument
Exception correctly thrown for wrong operand type passed to math operation (+,-,*,/) as first argument

Comparison functions:
10 ns > 10 ns = False
10 ns >= 10 ns = True
10 ns < 10 ns = False
10 ns <= 10 ns = True
10 ns == 10 ns = True
10 ns != 10 ns = False

10 ns > 15 ns = False
10 ns >= 15 ns = False
10 ns < 15 ns = True
10 ns <= 15 ns = True
10 ns == 15 ns = False
10 ns != 15 ns = True

Correctly got exception when comparing against invalid operand type as second argument
Correctly got exception when comparing against invalid operand type as first argument
Correctly got exception when checking equals against invalid operand type as second argument
Correctly got exception when checking equals against invalid operand type as first argument
Correctly got exception when checking not equals against invalid operand type as second argument
Correctly got exception when checking not equals against invalid operand type as first argument

Conversion to int:
'1 GHz' to long = 1000000000
'1 GHz'.getRoundedValue() = 1000000000

Conversion to float:
'1.77 s' to float = 1.770000
'1.77 s'.getFloatValue() = 1.770000

Conversion to float:
'1.77 s' to float = 1.770000
'1.77 s'.getFloatValue() = 1.770000

Conversion to bool:
bool(0 s) = False
bool(5 ns) = True

isValueZero():
'0 s'.isValueZero() = True
'5 ns'.isValueZero() = False

Negate:
-10 ns = -10 ns

Invert:
'10 ns'.invert() = 100 MHz

hasUnits():
'10 ns'.hasUnits("ns") = True
'10 ns'.hasUnits("s") = True
'10 ns'.hasUnits("Hz") = False
WARNING: No components are assigned to rank: 0.0
Simulation is complete, simulated time: 0 s
